### PR TITLE
CQ-4331591 Add Test for OOTB theme in AEM Project Archetype

### PR DESCRIPTION
**Pull request by suryansh29**
## Description

We needed to add testing for the content provided from the archetype. To test the themes provided by OOTB we decided to go with visual regression testing. The current plugins available for visual regression in wdio did not suit our requirement as they didn't support headless browsers. So we wrote a plugin for wdio based on pixel match library and are using that for image comparison.

The screenshots are added for firefox and chrome browsers on Mac, Linux, and Windows Operating systems. Total of 5(themes) x 3 (OS) x 2 (Browsers) x 7 (panels) = 210 screenshots.
They are stored in directory ui.tests/assets/form/themes/{theme name}/{OS}/{browser}/{screenshot}-base.png
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by pankaj-parashar** [comment]
@suryansh29 , As these themes are specific to forms, and assets folder is common for all the solutions in aem, I would recommend to move it to /assets/forms-theme folder for better identification. We need to capture all panels  as different forms component are part of different panels, and can't be captured in the single screen shot.

**Comment by catalan-adobe** [comment]
@suryansh29, the main goal of the `ui.tests` sub module is to demonstrate how customers could add UI tests to their project. The goal is not really to test the sample application or to test content (themese) provided into the Archetype.

If you want to test the provided themes you should add tests in the build part of the Archetype project.

Additional remarks:
* a more detailed description would help review process
* relying on image comparison might require more maintenance work. Who will ensure to update the base screenshots in the test module when needed?

**Comment by pankaj-parashar** [comment]
> @suryansh29, the main goal of the `ui.tests` sub module is to demonstrate how customers could add UI tests to their project. The goal is not really to test the sample application or to test content (themese) provided into the Archetype.
> 
> If you want to test the provided themes you should add tests in the build part of the Archetype project.
> 
> Additional remarks:
> 
> * a more detailed description would help review process
> * relying on image comparison might require more maintenance work. Who will ensure to update the base screenshots in the test module when needed?

@catalan-adobe , theme is visual part  and has to be deployed in runtime, AEM CS in this case. So these tests are part of ui.tests module which are executed in CM pipeline. I don't think adding this test as part of build step is feasible. Yes, we understand that image comparison is a maintainability challenge, that's why we have limited the combination of image viewport/os/browser etc. The purpose here is to just ensure that there is no visual regression for themes. In case there is intentional change to theme, the person making the change is responsible to update the screen shots. We will define the process for the same.

cc : @rismehta , @sakshi-arora1

**Comment by catalan-adobe** [comment]
Thanks for the clarification @suryansh29. Themes are new to me in AEM. Are they defined only in the Archetype?
Is there a dedicated project for these themes?

**Comment by pankaj-parashar** [comment]
> Thanks for the clarification @suryansh29. Themes are new to me in AEM. Are they defined only in the Archetype?
> Is there a dedicated project for these themes?

Themes are coming from AEM Forms. For creating a Form, a theme and template is prerequisites artifacts. When a customer is onboarded to AEM Forms CS, user need to create these artifacts before creating a form. To solve this issue and faster TTV, we are providing two templates and one theme already as part of archetype. Now we are adding few more themes as ootb content in this archetype. So all these artifacts are part of archetype. There is no separate project for these themes. Earlier, these artifacts were maintained in an internal repo, now we are moving these to this archetype repo. Ask/Goal here is that no theme should be added to archetype repo without test.

cc : @rismehta , @sakshi-arora1

**Comment by catalan-adobe** [comment]
Thanks again for clarification. 
I think these tests are useful for your team to check that the themes you contribute are working but I don't think these tests are useful for the users that will generate an application with it.
I still think that the testing of the themes contributed to the AEM Archetype should be handled in another flow.

Also, to be transparent, my only contribution to that repository is the `ui.tests` submodule. I would like to get further reviewing from the team owning the repository

**Comment by pankaj-parashar** [comment]
@catalan-adobe , thanks for your feedback. 
We discussed internally and agreed upon to add these tests here. By default these tests are not enabled for customer as their is opt-in mechanism for enabling these tests. So these are not going to execute in default deployment. If customer wants to refer these tests for writing their own test cases for their custom theme, this would be good starting point. So we are able to fulfil the following objectives here.

1. Validating the ootb content we are providing
2. Providing reference to customer for writing their own test case.
3. If customer uses ootb theme instead of creating their own, they can enable the test execution to keep these themes sanitised in their code base.

cc : @rismehta , @sakshi-arora1

File: src/main/archetype/ui.tests/test-module/specs/aem/forms.js [hunk 1]
```diff
@@ -220,4 +221,21 @@ describe('AEM Forms OOTB Content Tests', () => {
...
+            browser.AEMLogin(config.aem.author.username, config.aem.author.password);
+        });
+
+        const getThemePath = (theme) => '/content/dam/formsanddocuments-themes/${appId}/'+theme+'/jcr:content?wcmmode=disabled';
+        const themes = ['beryl', 'ultramarine', 'urbane', 'tranquil', 'canvas-3-0'];
+        themes.forEach((theme) => {
+            it(theme, () => {
+                browser.url(getThemePath(theme));
+                const isMatched = browser.call(() => browser.matchScreenshot('Theme preview', { errorThreshold: .2 }));
+                expect(isMatched, 'Screenshot did not match').true;
```
Comment (rismehta) [inline src/main/archetype/ui.tests/test-module/specs/aem/forms.js:236]: Can we improve this message by mentioning the theme page in the msg ?

File: src/main/archetype/ui.tests/test-module/specs/aem/forms.js [hunk 2]
```diff
@@ -220,4 +221,21 @@ describe('AEM Forms OOTB Content Tests', () => {
 
     });
 
+    describe('Theme testing', () => {
```
Comment (rismehta) [inline src/main/archetype/ui.tests/test-module/specs/aem/forms.js:224]: Are we removing the test case, if the `forms` addon is not enabled in the archetype ?

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable no-unused-vars */
+const PNG = require('pngjs').PNG;
+const pixelmatch = require('pixelmatch');
+const logger = require('@wdio/logger').default;
+const log = logger('PixelMatchPlugin');
+const { checkFileExists, makeDirectories, readPng, writePng } = require('./fileutils');
+const comparisonStrategy = require('./comparisonstrategy');
+const PathUtils = require('./pathutils');
+class PixelMatchPlugin {
```
Comment (rismehta) [inline src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:9]: Looks like the documentation is incomplete. Could you rectify this ?

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable no-unused-vars */
```
Comment (rismehta) [inline src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:1]: Do we need to add copyright headers here ?

File: src/main/archetype/ui.tests/docker-compose-wdio-chrome.yaml [hunk 5]
```diff
@@ -34,6 +34,7 @@ services:
     volumes:
       - ./target/reports:/tmp/reports
       - 'shared:/tmp/shared'
+      - 'shared:/usr/src/app/assets'
```
Comment (rismehta) [inline src/main/archetype/ui.tests/docker-compose-wdio-chrome.yaml:37]: I assume this is to compare the test image resource during the run, I don't see the code which copies the source image to this directory ?

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+const logger = require('@wdio/logger').default;
```
Comment (rismehta) [inline src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js:1]: Can we place the png images inside a folder name which is meaningful ? For example, inside a top level namespace  `forms/themes`

**Review by rismehta — CHANGES_REQUESTED** [review]

File: src/main/archetype/ui.tests/docker-compose-wdio-chrome.yaml (hunk 5 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/docker-compose-wdio-chrome.yaml:37]: This is the configuration for docker volume, what it does is the volume ui.tests will have files from the folder specified here.
```
      - 'shared:/usr/src/app/assets'
```
The docker will map this folder from container to volume, the volume will have persistent data.
ie the docker handles the copying.

File: src/main/archetype/ui.tests/test-module/specs/aem/forms.js (hunk 2 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/specs/aem/forms.js:224]: @pankaj-parashar  can you comment on this?

Comment (pankaj-parashar) [reply src/main/archetype/ui.tests/test-module/specs/aem/forms.js:224]: @rismehta , that has been done already when we added test for templates and fdm.
cc : @suryansh29

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js (hunk 6 above)
Comment (pankaj-parashar) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js:1]: @suryansh, you can move images as mentioned in https://github.com/adobe/aem-project-archetype/pull/802#issuecomment-919670141

**Review by pankaj-parashar — CHANGES_REQUESTED** [review]

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js (hunk 4 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:1]: I am not sure whether open source project should have copyright header, @pankaj-parashar  can confirm

Comment (pankaj-parashar) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:1]: @suryansh29 , please add copyright header to all js file added by us. please refer copyright header at https://github.com/adobe/aem-project-archetype/blob/develop/src/main/archetype/ui.tests/test-module/specs/aem/forms.js

File: src/main/archetype/ui.tests/test-module/specs/aem/forms.js (hunk 1 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/specs/aem/forms.js:236]: done

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js (hunk 3 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:9]: Documentation added

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js (hunk 4 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/launcher.js:1]: added

File: src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js (hunk 6 above)
Comment (suryansh29) [reply src/main/archetype/ui.tests/test-module/lib/wdio-pixelmatch-service/comparisonstrategy.js:1]: done

**Comment by suryansh29** [comment]
@pankaj-parashar  @rismehta  Changes done, added screenshots for windows, mac and linux and chrome firefox combinations.

**Comment by suryansh29** [comment]
@pankaj-parashar Haven't run these integration tests on CI yet, how can we run these tests on CI before merging?

**Review by pankaj-parashar — COMMENTED** [review]
> merging



> @pankaj-parashar Haven't run these integration tests on CI yet, how can we run these tests on CI before merging?

In this repo, pipeline execution (Deploy to CM Stage) happens post merging of PR.

**Review by rismehta — APPROVED** [review]

File: src/main/archetype/ui.tests/test-module/specs/aem/forms.js [hunk 7]
```diff
@@ -220,4 +221,21 @@ describe('AEM Forms OOTB Content Tests', () => {
...
+        before(() => {
+            browser.url(config.aem.author.base_url);
+            browser.AEMLogin(config.aem.author.username, config.aem.author.password);
```
[the remark this answers is not in the document]
Comment (rismehta) [reply src/main/archetype/ui.tests/test-module/specs/aem/forms.js]: Please address this review comment

**Pull request merged**